fontforgeexe/pythonui.c

```diff
@@ -50,41 +51,215 @@
+ * the convention the translator is using).
+ *
+ * ti.text is a (unichar_t *) but can be assigned a utf8 string when setting
+ * ti.text_is_1byte to true. ti.text_untranslated is a (char *) that we just
+ * have to hope can handle non-ASCII UTF-8 content, because that may be all we
+ * have.
+ */
+
+static const char mn_order[] = "}{:[]&=~%^;?+/,#$-@*!.7869054321QXZVJWFYGKUBHPCDMLTNSIROEA";
+
+struct py_menu_item {
     PyObject *func;
-    PyObject *check_enabled;		/* May be None (which I change to NULL) */
+    PyObject *check;		/* May be None (which I change to NULL) */
     PyObject *data;			/* May be None (left as None) */
-} *cvpy_menu_data = NULL, *fvpy_menu_data = NULL;
-static int cvpy_menu_cnt=0, cvpy_menu_max = 0;
-static int fvpy_menu_cnt=0, fvpy_menu_max = 0;
+};
+
+typedef void (*ff_menu_callback)(GWindow gw, struct gmenuitem *mi, GEvent *e);
+
+enum py_menu_type { pmt_font=0, pmt_char=1, pmt_size=2 };
+enum py_menu_flag { pmf_font=1, pmf_char=2 };
+
+static struct py_menu_data {
+    struct py_menu_item *items;
+    int cnt, max;
+    GMenuItem2 *menu;
+    ff_menu_callback moveto, invoke;
+    void (*setmenu)(GMenuItem2 *menu);
+    uint8 mn_offset;
+    uint8 mn_avail[128];
+    char *hotkey_prefix;
+} *py_menus;
 
-GMenuItem2 *cvpy_menu, *fvpy_menu;
+static unichar_t AllocateNextMnemonic(struct py_menu_data *pmd) {
+    while ( *(mn_order+pmd->mn_offset)!=0 && !pmd->mn_avail[*(mn_order + pmd->mn_offset)] )
+	++pmd->mn_offset;
+
+    char r = *(mn_order+pmd->mn_offset);
+    if ( r!=0 )
+	pmd->mn_avail[r] = false;
+
+    return (unichar_t) r;
+}
+
+static unichar_t FindMnemonic(const char *menu_string, unichar_t alt, struct py_menu_data *pmd) {
+    uint8 *cp;
+
+    if ( pmd->mn_avail[alt] ) {
+	pmd->mn_avail[alt] = false;
+	return alt;
+    }
+    for ( cp = (uint8 *) menu_string; *cp!=0; ++cp ) {
+	if ( *cp > 127 )
+	    continue;
+	alt = (unichar_t) toupper(*cp);
+	if ( pmd->mn_avail[alt] ) {
+	    pmd->mn_avail[alt] = false;
+	    return alt;
+	}
+    }
+    return 0;
+}
+
+static char *mncopy(const char *str, unichar_t *mn, int extra) {
+    unichar_t t=0;
+    const char *src = str;
+    char *dst, *ret;
+    size_t l=0;
+
+    if ( str==NULL ) {
+	if ( mn!=NULL )
+	    *mn = 0;
+	return NULL;
+    }
+
+    while ( *src!=0 ) {
+	if ( *src!='_' )
+	    ++l;
+	++src;
+    }
+
+    src = str;
+    ret = dst = malloc(l+1+extra);
+    for( ; *src; src++ ) {
+        if( *src == '_' ) {
+	    t = *(src+1);
+            continue;
+	}
+        *dst = *src;
+        dst++;
+    }
+    *dst = '\0';
+
+    if ( mn!=NULL )
+	*mn = t;
+    return ret;
+}
 
-static void py_tllistcheck(struct gmenuitem *mi,PyObject *owner,
-	struct python_menu_info *menu_data, int menu_cnt) {
+// Make sure there is no suffix when c==0 and exactly one suffix " (c)" when c!=0
+static char *SetMnemonicSuffix(const char*menu_string, unichar_t c) {
+    char *r = mncopy(menu_string, NULL, 4);
+    int l = strlen(r), i=l-1;
+
```

Comment:
Are you sure about the bounds-checking here? We can't count on a null terminator going backwards. If `r` is "", it seems like we might get a segmentation fault checking `*(r+i)` with `i = -1`. `r = " "` and other edge cases also seem likely to cause problems.

Reply:
I rewrote this a few times and am trying to remember why it wound up this way.

Clearly if menu_string is NULL you're already in trouble, so I guess I should return an `extra`-length zeroed out string in that case. That would also make `strlen()` happy. 

For the rest of it could I just add `i>=0 && ` to all the guards before the last one?

Reply:
I improved this some. `registerMenuItem()` rejects null name strings but I confirmed this works with empty strings and does something appropriate with all-space strings and trailing strings. It also passes my old test sample.